### ISSUE TITLE
If I modify the files near the end of my script, they don't get uploaded.

### ISSUE THREAD

**Issue by lukas** [wandb/client#24]
Version 0.4.17

If I modify the files near the end of my script, they don't get uploaded.

**Comment by shawnlewis** [comment]
More info: putting a time.sleep(10) before the script ends allow the files to be uploaded

**Comment by vanpelt** [comment]
@lukas can you do a couple runs today and confirm @shawnlewis's fix works?

**Comment by shawnlewis** [comment]
Even if it does, I think we should leave the issue open. Or create a new issue "Ensure that files are always uploaded" that references this one.

**Comment by shawnlewis** [comment]
I have a theory for this. Pushing files happens on the watchdog thread. When the main script exits, we wait 10s then tell the watchdog thread to stop. If the watchdog thread is uploading a big file that takes longer than 10s, it'll only get to upload that one file before we tell it stop. If there are other files pending after that, they'll get dropped.

This is just a theory, haven't confirmed. But it seems reasonable.

Ideally the watchdog thread would just serve notifications to another thread that manages pushing files. The pusher thread would wait until all file modifications are saved before exiting.

Even more ideally we'd let the user know what's going on. Maybe after their script exits we enable more verbose logging. Or at the end of the script we print "syncing remaining files, please monitor "http://app.wandb.ai/.../run" for sync progress. And then show good progress in the UI. Or a combination of the two.

**Comment by vanpelt** [comment]
I agree we should throw the pushes onto a another thread, and they should ideally be able to happen in parallel.  I had thought watchdog handled calling handlers in parallel, we should confirm that a long running event handler blocks future event handlers from firing.

**Comment by shawnlewis** [comment]
Here's a gist that shows the behavior: https://gist.github.com/shawnlewis/15c6903d3a653b6166e36ef4763ba30a

**Comment by shawnlewis** [comment]
This is fixed now, with FilePusher, deferring saves til the end, and not ignoring 0-length file creations.
